neps/nep-0621.md

```diff
@@ -0,0 +1,491 @@
+pub struct VaultDeposit {
+    /// The account that sends the deposit (payer of the assets).
+    pub sender_id: AccountId,
+
+    /// The account that receives the minted shares.
+    pub owner_id: AccountId,
+
+    /// Amount of underlying assets deposited into the vault.
+    pub assets: U128,
+
+    /// Amount of shares minted and issued to `owner_id`.
+    pub shares: U128,
+
+    /// Optional memo provided by the sender for off-chain use.
+    pub memo: Option<String>,
+}
+
+/// Event emitted when shares are redeemed from the vault.
+///
+/// Upon redemption, the vault burns the shares from `owner_id`
+/// and transfers the equivalent assets to `receiver_id`.
+pub struct VaultWithdraw {
+    /// The account that owns the shares being redeemed (burned).
+    pub owner_id: AccountId,
+
+    /// The account receiving the underlying assets.
+    pub receiver_id: AccountId,
+
+    /// Amount of shares redeemed (burned from the vault).
+    pub shares: U128,
+
+    /// Amount of underlying assets withdrawn from the vault.
+    pub assets: U128,
+
+    /// Optional memo provided by the redeemer for off-chain use.
+    pub memo: Option<String>,
+}
+```
+
+Deposit and Mint
+
+````rust
+pub trait FungibleTokenReceiver {
+    /// According to the NEP-141 standard, all fungible token transfers to a contract
+    /// must invoke the `ft_on_transfer` callback.
+    /// In the context of ERC-4626, both the **`deposit`** and **`mint`** functions
+    /// are implemented through this entrypoint on the vault contract.
+    ///
+    /// While NEP-141 does not prescribe how the `msg` parameter should be structured,
+    /// our suggested convention is to pass a JSON-encoded message with the following schema:
+    ///
+    /// ```rust
+    /// pub struct DepositMessage {
+    ///     /// The minimum number of shares that must be received for this deposit to succeed. (Slippage control)
+    ///     min_shares: Option<U128>,
+    ///
+    ///     /// The maximum number of shares that can be minted, used for `mint` operations.
+    ///     max_shares: Option<U128>,
+    ///
+    ///     /// The account that should receive the minted shares.
+    ///     receiver_id: Option<AccountId>,
+    ///
+    ///     /// Optional memo for logging or off-chain indexing.
+    ///     memo: Option<String>,
+    ///
+    ///     /// If `true`, the transfer is treated as a donation and no shares are minted.
+    ///     donate: Option<bool>,
+    /// }
+    /// ```
+    ///
+    /// Implementers are free to define their own schema for `msg`, but the vault must
+    /// correctly handle `ft_on_transfer` according to NEP-141:
+    /// - Return `PromiseOrValue::Value(0)` if all tokens were accepted.
+    /// - Return a non-zero `U128` to indicate the number of tokens to refund.
+    fn ft_on_transfer(&mut self, sender_id: AccountId, amount: U128, msg: String) -> PromiseOrValue<U128>
+}
+````
+
+```rust
+pub trait VaultOptionalSlippageWithdraw {
```

Comment:
are these two new methods mandatory in the NEP or just recommended? 

I see they add a very valid use case but they are not part of ERC4626